src/python/deepgnn/graph_engine/snark/tests/convert_test.py

```diff
@@ -722,6 +776,386 @@ def test_sanity_edge_sparse_features_data(graph_with_sparse_features):
+        )
+        npt.assert_equal(np.frombuffer(result[24:40], dtype=np.uint64), [3, 4])
+        npt.assert_almost_equal(
+            np.frombuffer(result[40:48], dtype=np.float32), [3.3, 4.4]
+        )
+        npt.assert_equal(np.frombuffer(result[48:64], dtype=np.uint64), [5, 6])
+        npt.assert_almost_equal(
+            np.frombuffer(result[64:72], dtype=np.float32), [5.5, 6.6]
+        )
+
+    output = tempfile.TemporaryDirectory()
+    data_data = [
+        "0,-1,uint64,2,1,2,1.1,2.2,int32,2,1,2\n",
+        "1,-1,uint64,2,3,4,3.3,4.4,int32,2,3,4\n",
+        "2,-1,uint64,2,5,6,5.5,6.6,int32,2,5,6\n",
+    ]
+    meta_data = {
+        "default_node_type": 0,
+        "default_node_weight": 1.5,
+        "default_node_feature_types": [None, "float32"],
+        "default_node_feature_lens": [None, [2]],
+    }
+    _gen_edge_list(output, data_data, meta_data)
+    with open("{}/node_features_{}_{}.data".format(output.name, 0, 0), "rb") as nfd:
+        expected_size = 96
+        result = nfd.read(expected_size + 1)
+        assert len(result) == expected_size
+        npt.assert_equal(np.frombuffer(result[0:16], dtype=np.uint64), [1, 2])
+        npt.assert_almost_equal(
+            np.frombuffer(result[16:24], dtype=np.float32), [1.1, 2.2]
+        )
+        npt.assert_equal(np.frombuffer(result[24:32], dtype=np.int32), [1, 2])
+        npt.assert_equal(np.frombuffer(result[32:48], dtype=np.uint64), [3, 4])
+        npt.assert_almost_equal(
+            np.frombuffer(result[48:56], dtype=np.float32), [3.3, 4.4]
+        )
+        npt.assert_equal(np.frombuffer(result[56:64], dtype=np.int32), [3, 4])
+        npt.assert_equal(np.frombuffer(result[64:80], dtype=np.uint64), [5, 6])
+        npt.assert_almost_equal(
+            np.frombuffer(result[80:88], dtype=np.float32), [5.5, 6.6]
+        )
+        npt.assert_equal(np.frombuffer(result[88:96], dtype=np.int32), [5, 6])
+
+
+def test_edge_list_header_multiple_partitions():
+    output = tempfile.TemporaryDirectory()
+    data_data = [
+        "0,-1\n",
+        "1,-1\n",
+        "2,-1\n",
+    ]
+    meta_data = {"default_node_type": 0, "default_node_weight": 1.5}
+    _gen_edge_list(output, data_data, meta_data, partitions=2)
+    with open("{}/node_{}_{}.map".format(output.name, 0, 0), "rb") as nm:
+        expected_size = 2 * (2 * 8 + 4)
+        result = nm.read(expected_size + 8)
+        assert len(result) == expected_size
+        assert result[0:8] == (0).to_bytes(8, byteorder=sys.byteorder)
+        assert result[8:16] == (0).to_bytes(8, byteorder=sys.byteorder)
+        assert result[16:20] == (0).to_bytes(4, byteorder=sys.byteorder)
+        assert result[20:28] == (2).to_bytes(8, byteorder=sys.byteorder)
+        assert result[28:36] == (1).to_bytes(8, byteorder=sys.byteorder)
+        assert result[36:40] == (0).to_bytes(4, byteorder=sys.byteorder)
+
+
+def test_edge_list_error_checking():
```

Comment:
where are tests for sparse feature parsing?

Reply:
All of the normal ones, its in the pytest.parameterize: https://github.com/microsoft/DeepGNN/blob/cdieckhaus/externalize_linear/src/python/deepgnn/graph_engine/snark/tests/convert_test.py#:~:text=def%20graph_with_sparse_features(request)%3A,output%20%3D%20tempfile.TemporaryDirectory()

Reply:
These are not parsing tests, but functional tests to verify a correct conversion logic across formats.
For parsing tests you need to handle various inputs for EdgeList only format, similarly what you did for binary features.

Reply:
Why do we only need this for EdgeList format? EdgeList has all of the same sparse feature tests as Json and TSV.

Reply:
Because json format uses standard library to parse json strings, but EdgeList format is using a hand-made parser. TSV doesn't support sparse features as I recall.

Reply:
I've added these.